### PR TITLE
[Kernel fusion] QK Norm + RoPE + Cache + Quant  

### PR DESCRIPTION
**Pull request by EricccYang**
## Purpose

  Resolves https://github.com/vllm-project/vllm/issues/36066                                                                                                           
  
  For models that use QK RMSNorm (e.g. Qwen3), the post-projection path currently requires                                                                             
  multiple kernel launches:                                       
                                                                                                                                                                       
  1.  A triton kernel for Qknorm                                                                                                            
  2. A Triton kernel for RoPE
  3. `reshape_and_cache_flash` for KV cache write + optional FP8 quant                                                                                                 
                                                                                                                                                                       
  This PR fuses all of them into a **single CUDA kernel** (`fused_qk_norm_rope_cache_quant`),                                                                          
  eliminating intermediate global memory round-trips and multiple kernel launch overheads.   wo
   kernels and the extra kernel launch overhead.                                                                                                                       
                                                                                                                                                                       
  ## What's included                                                                                                                                                 

  **CUDA kernel** (`csrc/fused_qk_norm_rope_cache_quant.cu`)                                                                                                           
  - Warp-per-head design: one warp handles RMSNorm + RoPE + cache write for a single head
  - Supports both NeoX and GPT-J RoPE styles                                                                                                                           
  - Supports BF16/FP16 compute with optional FP8 KV cache quantization                                                                                               
  - Vectorized loads (vec2/vec4) and warp-shuffle RMSNorm (no shared memory)                                                                                           
  - V is written to cache first (fire-and-forget) while K normalization runs, hiding store latency                                                                     
                                                                                                                                                                       
  **Compiler fusion pass** (`vllm/compilation/passes/fusion/qk_norm_rope_cache_quant_fusion.py`)                                                                       
  - Pattern-matches qknorm+rope+optional quant+cache write in the Inductor post-grad graph                                                                   
  - Replaces them with a single fused op call, resolving runtime KV cache context via `layer_name`                                                                     
  - Gated behind `CompilationConfig.fuse_qk_norm_rope_cache_quant` (default off)                                                                                       
                                                                                                                                                                       
  **Tests & benchmarks**                                                                                                                                               
  - Standalone correctness test against the unfused baseline (`tests/kernels/test_fused_qk_norm_rope_cache_quant.py`)                                                  
  - Kernel-level benchmark: fused vs baseline (`tests/kernels/bench_fused_qk_norm_rope_cache_quant.py`)                                                                
  - NCU profiling harness (`tests/kernels/bench_fused_qknrc_ncu.py`)                                                                                                   
                                                                                                                                                                       
  ## Test Plan                                                                                                                                                         
                                                                                                                                                                       
  ```bash                                                                                                                                                            
  # Unit test (requires GPU)
  python -m pytest tests/kernels/test_fused_qk_norm_rope_cache_quant.py -v                                                                                             
  
  # Benchmark (fused vs baseline, various batch sizes)                                                                                                                 
  python tests/kernels/bench_fused_qk_norm_rope_cache_quant.py           


 # nsys



# E2E data
```
                                                                                            
                                                                                                                                                                       
   ## Test Result

### Nsys 

Rough numbers on H20, Qwen3-4B config (32 Q heads, 8 KV heads, head_dim=128):

| Batch | Unfused | Fused | Speedup |                                                                                                                                
  |------:|--------:|------:|--------:|                                                                                                                                
  |     small batch|  5.0us  | 2.3us |  2.17x  |                                                                                                                                
  |    middle batch | 17.0us  | 12.0us |  1.42x  |                                                                                                                               
  |   large batch | 28.2us  | 23.0us |  1.23x  |   



Unfused Max Cost 
<img width="2990" height="1914" alt="image" src="https://github.com/user-attachments/assets/6e824085-19aa-464f-80c1-58f8c60c4bc8" />

Unfused Min Cost 
<img width="2990" height="1914" alt="image" src="https://github.com/user-attachments/assets/e3241d6a-5780-46b5-bdf4-6be524cf7a72" />
                                                     
Fused Max Cost
<img width="2990" height="1914" alt="image" src="https://github.com/user-attachments/assets/a77e3fc5-9130-4fd4-a355-ec533bdd728f" />

Fuse Min Cost
<img width="2990" height="1914" alt="image" src="https://github.com/user-attachments/assets/0a8a5b8e-009d-4a09-9e83-1e3cd2738d80" />

                                                                                                   
                                                                                                                                                                       
                                                                         
                                                                                                                                                                     
   ## How to enable                                                                                                                                                        
                                                                                                                                                 
  Not enabled by default. To opt in, set in your vLLM config:                                                                                                          
  compilation_config={"pass_config": {"fuse_qk_norm_rope_cache_quant": True}}
                                                                                                                                                                                                                             
                                                                                                                                                                       
  ---                                                                                                                                                                  
                                                      
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Review by ProExpertProg — COMMENTED** [review]
Looking forward to this, ping when ready for review!

File: vllm/compilation/passes/fusion/qk_norm_rope_cache_quant_fusion.py [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+    num_heads = query.shape[1]
+    num_kv_heads = key.shape[1]
+    head_dim = query.shape[2]
+
+    # ---- Step 1: Q RMSNorm (normalise each head independently) ----
+    q_norm = torch.empty_like(query)
+    torch.ops._C.rms_norm(
+        result=q_norm,
+        input=query,
+        weight=q_weight.view(head_dim),
```
[the remark this answers is not in the document]
Comment (EricccYang) [reply vllm/compilation/passes/fusion/qk_norm_rope_cache_quant_fusion.py]: pattern match would fail

File: vllm/compilation/passes/fusion/qk_norm_rope_cache_quant_fusion.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+        key_cache, value_cache = kv_cache.unbind(0)
+        reshape_and_cache_flash(
+            key,
+            value,
+            key_cache,
+            value_cache,
+            layer_slot_mapping,
+            attn_layer.impl.kv_cache_dtype,
+            attn_layer._k_scale,
+            attn_layer._v_scale,
```
[the remark this answers is not in the document]
Comment (EricccYang) [reply vllm/compilation/passes/fusion/qk_norm_rope_cache_quant_fusion.py]: I think they do have their default values.
```
def set_default_quant_scales(layer: nn.Module, register_buffer: bool = False) -> None:
    """Sets default quantization scales for the layer."""
    if register_buffer:
        layer.register_buffer("_k_scale", torch.tensor(1.0, dtype=torch.float32))
        layer.register_buffer("_v_scale", torch.tensor(1.0, dtype=torch.float32))
        layer.register_buffer("_q_scale", torch.tensor(1.0, dtype=torch.float32))
        layer.register_buffer("_prob_scale", torch.tensor(1.0, dtype=torch.float32))
    else:
        layer._k_scale.fill_(1.0)
        layer._v_scale.fill_(1.0)
        layer._q_scale.fill_(1.0)
        layer._prob_scale.fill_(1.0)
```

File: csrc/fused_qk_norm_rope_cache_quant.cu [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+
+  // ── Token pointers ──
+  const scalar_t* q_in = qkv + (int64_t)token_idx * (q_size + 2 * kv_size);
+  const scalar_t* k_in = q_in + q_size;
+  const scalar_t* v_in = k_in + kv_size;
+
+  // ── Cos / Sin for this position ──
+  const int64_t pos = positions[token_idx];
+  const scalar_t* cos_ptr = cos_sin_cache + pos * head_dim;
+  const scalar_t* sin_ptr = cos_ptr + embed_dim;
```
[the remark this answers is not in the document]
Comment (EricccYang) [reply csrc/fused_qk_norm_rope_cache_quant.cu]: updated

**Comment by zhao18235158400** [comment]
Hi @EricccYang — I read your QK Norm + RoPE + KV cache + optional FP8 quant fusion work closely, especially the single CUDA-kernel design, the NCU harness, and the H20 measurements showing up to 2.17× kernel-level speedup at small batch sizes. I’m Daniel, a user researcher with a product research team exploring a developer-owned local AI system for engineers who modify inference runtimes and kernels. We’re trying to understand what would make work like this easier across devices: hardware access, stable low-level APIs, profiler visibility, reproducible kernel/e2e benchmarks, or something else. Would you be open to a Zoom conversation of up to 30 minutes? We’re at the prototype stage and may later provide test units or invite relevant participants as early users or technical advisors. This is research, not sales. If interested, what private channel would you prefer?
